### PR TITLE
LevelsScreenTest

### PR DESCRIPTION
**Pull request by mlytvynchuk**
Covered the levels screen with tests.

### CODE REVIEW

File: app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
+package com.google.droidjump;
```
Comment (redls) [inline app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java:1]: For tests, we try and maintain the following structure (basically each test has 3 blocks of code):
{
// setup of the test

// action which needs testing 

// checking that the behaviour obtained is expected
}

Comment (mlytvynchuk) [reply app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java:1]: Fixed

File: app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+        assertEquals(LevelManager.getCurrentLevel(), lastLevel);
+        assertEquals(LevelManager.getLastLevel(), lastLevel);
+    }
+
+    @Test
+    public void chooseAvailableLevel() {
+        int lastLevel = LevelManager.getLastLevel();
+        // Available level is the lever from range [firstLevel, lastLevel].
+        int availableLevel = (int) (Math.random() *
```
Comment (saimonsaret) [inline app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java:61]: Using random values in tests is a bad pattern and can lead to flakiness. You can choose any predetermined values here.

File: app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    public void chooseNotAvailableLevel() {
+        int levelsCount = LevelManager.getLevelsCount();
+        if (levelsCount > 1) {
+            int lastLevel = LevelManager.getLastLevel();
+            if (lastLevel == levelsCount) {
+                LevelManager.resetGameData();
+                lastLevel = LevelManager.getLastLevel();
+            }
+            int notAvailableLevelStartsFromLevel = lastLevel + 1;
+            int notAvailableLevel = (int) (Math.random() *
```
Comment (saimonsaret) [inline app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java:79]: Same

File: app/src/androidTest/java/com/google/droidjump/StartScreenTest.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    public void navigateToLevelsScreen() {
+        onView(withId(R.id.level_button)).perform(click());
+
+        onView(withId(R.id.choose_level_header)).check(matches(isDisplayed()));
+    }
+
+    @Test
+    public void startNewGame() {
+        // Set the handpicked current level.
+        int testingCurrentLevel = GameConstants.GAME_LEVELS_COUNT / 2;
```
Comment (saimonsaret) [inline app/src/androidTest/java/com/google/droidjump/StartScreenTest.java:63]: Test would change if new levels added. Please use a constant value here.

File: app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    @Before
+    public void setUp() throws Exception {
+        // Running the levels screen.
+        activityTestRule.getActivity().getSupportFragmentManager()
+                .beginTransaction().replace(R.id.activity_wrapper, new LevelsFragment());
+        onView(withId(R.id.level_button)).perform(click());
+
+        // Setting the game data.
+        int lastLevel = (int) (Math.random() * (LevelManager.getLevelsCount()
```
Comment (saimonsaret) [inline app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java:34]: Same here about random values.

File: app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    @Test
+    public void checkIfHeaderIsDisplayed() {
+        onView(withId(R.id.choose_level_header)).check(matches(isDisplayed()));
+    }
+
+    @Test
+    public void chooseLastLevel() {
+        int lastLevelIndex = LevelManager.getLastLevelIndex();
+        String lastLevelName = LevelManager.getGameLevels().get(lastLevelIndex).getLevelName();
+        onView(withText(lastLevelName)).perform(ViewActions.click());
```
Comment (redls) [inline app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java:50]: add new line between these 2 lines

File: app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        assertEquals(LevelManager.getCurrentLevelIndex(), lastLevelIndex);
+        assertEquals(LevelManager.getLastLevelIndex(), lastLevelIndex);
+    }
+
+    @Test
+    public void chooseAvailableLevel() {
+        // Available level is the lever from range [firstLevel, lastLevel].
+        int availableLevelIndex = LevelManager.getCurrentLevelIndex();
+        String availableLevelName = LevelManager.getGameLevels().get(availableLevelIndex).getLevelName();
+        onView(withText(String.valueOf(availableLevelName))).perform(click());
```
Comment (redls) [inline app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java:61]: ditto

File: app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+        assertEquals(LevelManager.getCurrentLevelIndex(), availableLevelIndex);
+    }
+
+    @Test
+    public void chooseNotAvailableLevel() {
```
Comment (redls) [inline app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java:67]: I am not sure what this test checks?

Comment (mlytvynchuk) [reply app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java:67]: It checks the user can start to play not available level for them.

File: app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java (hunk 6 above)
Comment (mlytvynchuk) [reply app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java:50]: Added.

File: app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java (hunk 7 above)
Comment (mlytvynchuk) [reply app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java:61]: Added.

File: app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java (hunk 8 above)
Comment (mlytvynchuk) [reply app/src/androidTest/java/com/google/droidjump/LevelsScreenTest.java]: Renamed.

File: app/src/main/java/com/google/droidjump/models/LevelManager.java [hunk 9]
```diff
@@ -89,12 +87,11 @@ public static LevelStrategy getCurrentLevelStrategy() {
...
     }
 
-    public static String getCurrentLevelName(){
+    public static String getCurrentLevelName() {
         return gameLevels.get(getCurrentLevelIndex()).getLevelName();
     }
 
-    public static ArrayList<LevelConfig> getGameLevels(){
+    public static ArrayList<LevelConfig> getGameLevels() {
         return (ArrayList<LevelConfig>) gameLevels.clone();
```
Comment (redls) [inline app/src/main/java/com/google/droidjump/models/LevelManager.java:95]: We should just return List<LevelConfig>.

I am not sure why we are doing a .clone()? Why aren't we just returning the list?

Comment (mlytvynchuk) [reply app/src/main/java/com/google/droidjump/models/LevelManager.java:95]: Not my code..

Comment (saimonsaret) [reply app/src/main/java/com/google/droidjump/models/LevelManager.java:95]: All code is our team's code. We didn't catch it previously, but now that we have noticed it is better to fix it. 

If you don't want to fix it here as this is not what your PR is about, you can make a separate PR fixing this problem or ask Daria why it is like that - maybe it was actually intended. If not, you could ask her to fix it or fix yourself.

**Review by saimonsaret — APPROVED** [review]

File: app/src/main/java/com/google/droidjump/models/LevelManager.java (hunk 9 above)
Comment (mlytvynchuk) [reply app/src/main/java/com/google/droidjump/models/LevelManager.java:95]: We discussed with Daria that she will fix it.

**Pull request merged**